psa-crypto/src/types/key.rs

```diff
@@ -0,0 +1,768 @@
+    BrainpoolPR1,
+    /// Curve used primarily in France and elsewhere in Europe.
+    /// This family comprises one 256-bit curve:
+    ///   * FRP256v1: `bits` = 256
+    Frp,
+    /// Montgomery curves.
+    /// This family comprises the following Montgomery curves:
+    ///   * Curve25519: `bits` = 255
+    ///   * Curve448: `bits` = 448
+    Montgomery,
+}
+
+/// Enumeration of Diffie Hellman group families supported.
+#[derive(Copy, Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub enum DhFamily {
+    /// Diffie-Hellman groups defined in RFC 7919 Appendix A.
+    /// This family includes groups with the following `bits`: 2048, 3072, 4096, 6144, 8192.
+    /// An implementation can support all of these sizes or only a subset.
+    Rfc7919,
+}
+
+/// Definition of the key policy, what is permitted to do with the key.
+#[derive(Copy, Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct Policy {
+    /// Usage flags for the key.
+    pub usage_flags: UsageFlags,
+    /// Permitted algorithms to be used with the key.
+    pub permitted_algorithms: Algorithm,
+}
+
+/// Definition of the usage flags. They encode what kind of operations are permitted on the key.
+#[derive(Copy, Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct UsageFlags {
+    /// Permission to export the key.
+    pub export: bool,
+    /// Permission to copy the key.
+    pub copy: bool,
+    /// Permission for the implementation to cache the key.
+    pub cache: bool,
+    /// Permission to encrypt a message with the key.
+    pub encrypt: bool,
+    /// Permission to decrypt a message with the key.
+    pub decrypt: bool,
+    /// Permission to sign a message with the key.
+    pub sign_message: bool,
+    /// Permission to verify a message signature with the key.
+    pub verify_message: bool,
+    /// Permission to sign a message hash with the key.
+    pub sign_hash: bool,
+    /// Permission to verify a message hash with the key.
+    pub verify_hash: bool,
+    /// Permission to derive other keys from this key.
+    pub derive: bool,
+}
+
+/// Definition of the key ID.
+#[derive(Copy, Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct Id {
+    pub(crate) id: psa_crypto_sys::psa_key_id_t,
+    pub(crate) handle: Option<psa_crypto_sys::psa_key_handle_t>,
+}
+
+impl Id {
+    pub(crate) fn handle(self) -> Result<psa_crypto_sys::psa_key_handle_t> {
+        Ok(match self.handle {
+            Some(handle) => handle,
+            None => {
+                let mut handle = 0;
+                status_to_result(unsafe { psa_crypto_sys::psa_open_key(self.id, &mut handle) })?;
+
+                handle
+            }
+        })
+    }
+
+    pub(crate) fn close_handle(self, handle: psa_crypto_sys::psa_key_handle_t) -> Result<()> {
+        if self.handle.is_none() {
+            status_to_result(unsafe { psa_crypto_sys::psa_close_key(handle) })
+        } else {
+            Ok(())
+        }
+    }
```

Comment:
Oh yes you are right, it would make more sense to directly use the one in `self`. I think I went through several refactoring of this so maybe I mixed things up.

Reply:
Oh wait actually there was a reason: the `handle` value inside the `Id` structure is only for `Id` that were creating through the key creations operations. It is `None` if it was created directly from a key ID and does not have a handle assigned to it. For this case we open/close the key before/after each call.